### ISSUE TITLE
Consider bitwise state flag implementation

### ISSUE THREAD

**Issue by surchs** [metaneuro/annotation_tool#53]
Hey @jarmoza, sorry, looks like I missed #42 by a couple of minutes with this comment. So I am turning this into a separate issue:

I have been giving this a bit more of a think. I understand the point behind this now and I do agree it is an elegant solution. But I think based on what we need so far, I vote to go with boolean flags and come back to this if and when things get unwieldy. For context, we are using this bit field flag style in 2 places, for a total of 6 states:

- `categorization.vue`
  - STATE_NOCATEGORIES_PAINTED
  - STATE_ATLEASTONE_CATEGORY_PAINTED
- `index.vue`
  - `STATE_NOFILES_LOADED`
  - `STATE_TSVFILE_LOADED`
  - `STATE_JSONFILE_LOADED`
  - `STATE_BOTHFILES_LOADED`

This can be done by:
- `categorization.vue`
  - `STATE_CATEGORY_PAINTED` = oneof(True, False)
- `index.vue`
  - `this.$store.state.pageData.home.tsvFile != null`
  - `this.$store.state.pageData.home.jsonFile`
  - (in fact this is how we are using it in `categorization.vue` already

The main benefit here is legibility, and how easy it is for us to come back and understand what's going on in the future. When we find that this is not performant enough at some point, we know how to fix it then. 

Does that sound OK to you?

_Originally posted by @surchs in https://github.com/metaneuro/annotation_tool/pull/42#discussion_r808450012_

**Comment by jarmoza** [comment]
I think it's also important to think about how the state is being used here. The control flow logic from the flag system actually grants flexibility. If you look at the code, what's happening is this:

1. `changeToNewState` is called at page creation or when something important happens.
2. `changeToNewState` asks `determineState` to figure out the state of the page by looking at relevant store data
3. `determineState` produces this flag variable which contains the bitwise flags of the state of the app
4. That flag (consisting of all the interactions that have occurred which were saved in the store) is sent to `changeState` which then uses that to figure out what should be done as a result

So it's really more about guiding the page logic to the right set of behaviors. You can see it's already somewhat complex on the landing page by the fact that a tsv could be loaded, a json could be loaded or both could be loaded.

```
			changeState(p_state) {

				// 1. Trigger the behavior for the requested state change
				switch ( p_state ) {

					// A. Handle changes for when a tsv file has been loaded
					case this.possibleStates.STATE_TSVFILE_LOADED:
					case this.possibleStates.STATE_BOTHFILES_LOADED:
						this.changeState_TsvFileLoaded();
						break;

					// B. No interface changes necessary
					case this.possibleStates.STATE_JSONFILE_LOADED:
						break;

					// C. Handle changes for when no files are loaded
					default:
						this.changeState_NoFilesLoaded();
						break;
				}
			}
```

In other words, having the ability to get the state in a compact fashion temporarily (in this case the newState variable from determineState) allows us to separate behaviors more easily (and yes, more elegantly).

**Comment by jarmoza** [comment]
From my design perspective, I am also thinking about the possibility where page state becomes just a little more complex and just looking at a series of boolean flags becomes less legible and worse, confusing to code and debug.

Since this implementation is working currently, perhaps let's just move on to the annotation page and see the complexity of the conditions required?

**Comment by surchs** [comment]
OK, if you see this as a reasonable design for future complexity (thinking also of integrating with dandi, NIDM, et al) then I trust your expertise here - let's go with this. I also agree that focusing on the annotation page comes first.

**Comment by surchs** [comment]
With #66 merged now, we can take another look at this. I had suggested one way of having the state represented only in the global state in https://github.com/metaneuro/annotation_tool/pull/66/commits/a0e79d4fb4bd81448fa20f8bb14fc60ae0778e8c (although there is a lot of formatting fluff).

**Comment by surchs** [comment]
Addressed in #108
